server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -202,37 +228,54 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
+                    if (minNodeVersion.before(FULL_CONCURRENCY_VERSION)) {
+                        throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName,
+                                "cannot snapshot while a snapshot deletion is in-progress in [" + deletionsInProgress + "]");
+                    } else {
+                        readyToExecute = deletionsInProgress.getEntries().stream().noneMatch(entry ->
+                                entry.repository().equals(repositoryName) && entry.state() == SnapshotDeletionsInProgress.State.META_DATA);
+                    }
                 }
                 final RepositoryCleanupInProgress repositoryCleanupInProgress = currentState.custom(RepositoryCleanupInProgress.TYPE);
                 if (repositoryCleanupInProgress != null && repositoryCleanupInProgress.hasCleanupInProgress()) {
                     throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName,
                         "cannot snapshot while a repository cleanup is in-progress in [" + repositoryCleanupInProgress + "]");
                 }
-                SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
                 // Fail if there are any concurrently running snapshots. The only exception to this being a snapshot in INIT state from a
                 // previous master that we can simply ignore and remove from the cluster state because we would clean it up from the
                 // cluster state anyway in #applyClusterState.
-                if (snapshots != null && snapshots.entries().stream().anyMatch(entry -> entry.state() != State.INIT)) {
+                if (minNodeVersion.before(FULL_CONCURRENCY_VERSION) && snapshots != null
+                        && runningSnapshots.stream().anyMatch(entry -> entry.state() != State.INIT)) {
                     throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
                 }
                 // Store newSnapshot here to be processed in clusterStateProcessed
                 List<String> indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState,
                     request.indicesOptions(), request.indices()));
                 logger.trace("[{}][{}] creating snapshot for indices [{}]", repositoryName, snapshotName, indices);
 
-                final List<IndexId> indexIds = repositoryData.resolveNewIndices(indices);
+                final List<IndexId> indexIds = repositoryData.resolveNewIndices(
+                        indices, runningSnapshots.stream().filter(entry -> entry.repository().equals(repositoryName))
+                                .flatMap(entry -> entry.indices().stream()).distinct()
+                                .collect(Collectors.toMap(IndexId::getName, Function.identity())));
```

Comment:
> Do we have a test somewhere that ensures that we're not resolving to new indices for concurrent snaps? (which would always work, but give us less incrementality)

I don't think this would work. Remember, we're looking up index uuids (the repo uuids) by index name. If we were to cause collisions there with multiple `IndexId` for a single name, then we would a. trip assertions in `RepositoryData` when adding a snapshot to it (updated shard generations would conflict with existing `RepositoryData`) and b. fail repository consistency checks. We do have a bunch of tests that run two or more concurrent snapshots on an empty repo that cover this properly I think (it was one of the first things that blew up when I enabled concurrent snapshots :)).